x/datadeal/keeper/deal.go

```diff
@@ -230,3 +233,81 @@ func (k Keeper) GetAllDataSaleList(ctx sdk.Context) ([]types.DataSale, error) {
 
 	return dataSales, nil
 }
+
+func (k Keeper) VoteDataDelivery(ctx sdk.Context, vote *types.DataDeliveryVote, signature []byte) error {
+	if err := vote.ValidateBasic(); err != nil {
+		return sdkerrors.Wrap(types.ErrDataDeliveryVote, err.Error())
+	}
+	if !k.verifyVoteSignature(ctx, vote, signature) {
+		return sdkerrors.Wrap(oracletypes.ErrDetectionMaliciousBehavior, "")
+	}
+
+	// Check if the dataSale vote status
+	if err := k.validateDataDeliveryVote(ctx, vote); err != nil {
+		return sdkerrors.Wrap(types.ErrDataDeliveryVote, err.Error())
+	}
+
+	// When all validations pass, the vote is saved.
+	if err := k.SetDataDeliveryVote(ctx, vote); err != nil {
+		return sdkerrors.Wrap(types.ErrDataDeliveryVote, err.Error())
+	}
+
+	return nil
+}
+
+func (k Keeper) verifyVoteSignature(ctx sdk.Context, vote *types.DataDeliveryVote, signature []byte) bool {
+	voteBz := k.cdc.MustMarshal(vote)
+
+	oraclePubKeyBz := k.oracleKeeper.GetParams(ctx).MustDecodeOraclePublicKey()
+	return secp256k1.PubKey(oraclePubKeyBz).VerifySignature(voteBz, signature)
+}
+
+func (k Keeper) validateDataDeliveryVote(ctx sdk.Context, vote *types.DataDeliveryVote) error {
+	dataSale, err := k.GetDataSale(ctx, vote.VerifiableCid, vote.DealId)
+	if err != nil {
+		return err
+	}
+	if dataSale.Status != types.DATA_SALE_STATUS_DELIVERY_VOTING_PERIOD {
+		return fmt.Errorf("the current voted data's status is not 'DELIVERT_VOTING_PERIOD'")
+	}
+	return nil
+}
```

Comment:
I think it need to verify a oracle's status, too

Reply:
Thanks, I will add it 👍